### PR TITLE
audiohook.c: resolving the issue with audiohook both reading when packet loss on one side of the call

### PR DESCRIPTION
**Pull request by Tinet-mucw**
audiohook.c: resolving the issue with audiohook both reading when packet loss on one side of the call

When there is 0% packet loss on one side of the call and 15% packet loss on the other side, reading frame is often failed when reading direction_both audiohook. when read_factory available = 0, write_factory available = 320; i think write factory is usable read; because after reading one frame, there is still another frame that can be read together with the next read factory frame.

Resolves: #851

### CODE REVIEW

**Comment by Tinet-mucw** [comment]
cherry-pick-to: 22
cherry-pick-to: 21
cherry-pick-to: 20
cherry-pick-to: 18

**Review by gtjoseph — DISMISSED** [review]
Would you mind adding some detail to the commit message on how your change fixes the issue?

Also, usable_read and usable_write are being set using ast_slinfactory_available just a few lines above this.  Is there a reason you added the available check for 2 * samples where you did instead of modifying the criteria for usable_read and usable_write?  I'm not saying the way you did it was wrong, I'm just not understanding.

**Comment by Tinet-mucw** [comment]
code 1

```c
	/* If we want to provide only a read factory make sure we aren't waiting for other audio */
	if (usable_read && !usable_write && (ast_tvdiff_ms(ast_tvnow(), audiohook->write_time) < (samples/8)*2)) {
		ast_debug(3, "Write factory %p was pretty quick last time, waiting for them.\n", &audiohook->write_factory);
		return NULL;
	}

	/* If we want to provide only a write factory make sure we aren't waiting for other audio */
	if (usable_write && !usable_read && (ast_tvdiff_ms(ast_tvnow(), audiohook->read_time) < (samples/8)*2)) {
		ast_debug(3, "Read factory %p was pretty quick last time, waiting for them.\n", &audiohook->read_factory);
		return NULL;
	}

```

code 2
```shell
int ast_audiohook_write_frame(struct ast_audiohook *audiohook, enum ast_audiohook_direction direction, struct ast_frame *frame)
{
...
	if (ast_test_flag(audiohook, AST_AUDIOHOOK_TRIGGER_SYNC) && (our_factory_ms - other_factory_ms > AST_AUDIOHOOK_SYNC_TOLERANCE)) {
		ast_debug(1, "Flushing audiohook %p so it remains in sync\n", audiohook);
		ast_slinfactory_flush(factory);
		ast_slinfactory_flush(other_factory);
	}

	if (ast_test_flag(audiohook, AST_AUDIOHOOK_SMALL_QUEUE) && ((our_factory_ms > AST_AUDIOHOOK_SMALL_QUEUE_TOLERANCE) || (other_factory_ms > AST_AUDIOHOOK_SMALL_QUEUE_TOLERANCE))) {
		ast_debug(1, "Audiohook %p has stale audio in its factories. Flushing them both\n", audiohook);
		ast_slinfactory_flush(factory);
		ast_slinfactory_flush(other_factory);
	} else if ((our_factory_ms > AST_AUDIOHOOK_LONG_QUEUE_TOLERANCE) || (other_factory_ms > AST_AUDIOHOOK_LONG_QUEUE_TOLERANCE)) {
		ast_debug(1, "Audiohook %p has stale audio in its factories. Flushing them both\n", audiohook);
		ast_slinfactory_flush(factory);
		ast_slinfactory_flush(other_factory);
	}
```


In my mind, code 1 avoid discrepancies in the recording times for both sides, which could result in data from two channels being read in two separate instances. 

However, in an extreme scenario where one side of the call has a packet loss rate of 50% and the other side has a packet loss rate of 100%

0s 0020ms `read channel` and  `write channel` write audiohook； read_factory： 160，  write_factory： 160
0s 0025ms read audiohook both direction; read_factory:0 write_factory:0
0s 0040ms `read channel` and  `write channel` write audiohook； read_factory： 0(50% loss)，  write_factory： 160
0s 0045ms read audiohook both direction read fail( for above code 1); read_factory:0 write_factory:160

0s 0060ms `read channel` and  `write channel` write audiohook； read_factory： 160，  write_factory： 320
0s 0065ms read audiohook both direction; read_factory:0 write_factory:160
0s 0080ms `read channel` and  `write channel` write audiohook； read_factory： 0(50% loss)，  write_factory： 320
0s 0085ms read audiohook both direction read fail( for above code 1); read_factory:0 write_factory:320

0s 0100ms `read channel` and  `write channel` write audiohook； read_factory： 160，  write_factory： 480
0s 0105ms read audiohook both direction; read_factory:0 write_factory:320
0s 0120ms `read channel` and  `write channel` write audiohook； read_factory： 0(50% loss)，  write_factory： 480
0s 0125ms read audiohook both direction read fail( for above code 1); read_factory:0 write_factory:480

0s 0140ms `read channel` and  `write channel` write audiohook； read_factory： 160，  write_factory： 640
0s 0145ms read audiohook both direction; read_factory:0 write_factory:480
0s 0160ms `read channel` and  `write channel` write audiohook； read_factory： 0(50% loss)，  write_factory： 640
0s 0165ms read audiohook both direction read fail( for above code 1); read_factory:0 write_factory:640

0s 0180ms `read channel` and  `write channel` write audiohook； read_factory： 160，  write_factory： 800
0s 0185ms read audiohook both direction; read_factory:0 write_factory:640
0s 0200ms `read channel` and  `write channel` write audiohook； read_factory： 0(50% loss)，  write_factory： 800
0s 0205ms read audiohook both direction read fail( for above code 1); read_factory:0 write_factory:800

0s 0220ms `read channel` and  `write channel` write audiohook； read_factory： 160，  write_factory： 960
0s 0225ms read audiohook both direction; read_factory:0 write_factory:800
0s 0240ms `read channel` and  `write channel` write audiohook； read_factory： 0(50% loss)，  write_factory： 960
0s 0245ms read audiohook both direction read fail( for above code); read_factory:0 write_factory:960

0s 0260ms `read channel` and  `write channel` write audiohook； read_factory： 160，  write_factory： 1120
0s 0265ms read audiohook both direction; read_factory:0 write_factory:960
0s 0280ms `read channel` and  `write channel` write audiohook； read_factory： 0(50% loss)，  write_factory： 1120
0s 0285ms read audiohook both direction read fail( for above code); read_factory:0 write_factory:1120

0s 0300ms `read channel` and  `write channel` write audiohook； read_factory： 160，  write_factory： 1280
0s 0305ms read audiohook both direction; read_factory:0 write_factory:1120
0s 0320ms `read channel` and  `write channel` write audiohook； read_factory： 0(50% loss)，  write_factory： 1280
0s 0325ms read audiohook both direction read fail( for above code); read_factory:0 write_factory:1280

0s 0340ms `read channel` and  `write channel` write audiohook； read_factory： 160，  write_factory： 1440
0s 0345ms read audiohook both direction; read_factory:0 write_factory:1280
0s 0360ms `read channel` and  `write channel` write audiohook； read_factory： 0(50% loss)，  write_factory： 1440
0s 0365ms read audiohook both direction read fail( for above code); read_factory:0 write_factory:1440

0s 0380ms `read channel` and  `write channel` write audiohook； read_factory： 160，  write_factory： 1600
0s 0385ms read audiohook both direction; read_factory:0 write_factory:1440
0s 0400ms `read channel` and  `write channel` write audiohook； read_factory： 0(50% loss)，  write_factory： 1600
0s 0405ms read audiohook both direction read fail( for above code); read_factory:0 write_factory:1600

0s 0340ms `read channel` and  `write channel` write audiohook； read_factory： 160，  write_factory： 1760
0s 0345ms read audiohook both direction; read_factory:0 write_factory:1600
0s 0360ms `read channel` and  `write channel` write audiohook； read_factory： 0(50% loss)，  write_factory： 1760
0s 0365ms read audiohook both direction read fail( for above code); read_factory:0 write_factory:1760

0s 0380ms `read channel` and  `write channel` write audiohook；ast_slinfactory_flush(for code 2) read_factory： 160，  write_factory： 160
0s 0385ms read audiohook both direction; read_factory:0 write_factory:0
0s 0400ms `read channel` and  `write channel` write audiohook； read_factory： 0(50% loss)，  write_factory： 160
0s 0405ms read audiohook both direction read fail( for above code); read_factory:0 write_factory:160

Based on the above data, half of the frames fail to be read.


code 3
```c
	/* If we want to provide only a read factory make sure we aren't waiting for other audio */
	if (usable_read && !usable_write && (ast_tvdiff_ms(ast_tvnow(), audiohook->write_time) < (samples/8)*2) && (ast_slinfactory_available(&audiohook->read_factory) < 2 * samples)) {
		ast_debug(3, "Write factory %p was pretty quick last time, waiting for them.\n", &audiohook->write_factory);
		return NULL;
	}

	/* If we want to provide only a write factory make sure we aren't waiting for other audio */
	if (usable_write && !usable_read && (ast_tvdiff_ms(ast_tvnow(), audiohook->read_time) < (samples/8)*2) && (ast_slinfactory_available(&audiohook->write_factory) < 2 * samples)) {
		ast_debug(3, "Read factory %p was pretty quick last time, waiting for them.\n", &audiohook->read_factory);
		
```

when read_factory available = 0, write_factory available = 320; i think write factory is usable read; 
because after reading one frame, there is still another frame that can be read together with the next read factory frame.



My English language skills are limited; could you help me rewrite the commit? and i will change it later

**Comment by Tinet-mucw** [comment]
The assignment of usable_write needs to comprehensively consider the read factory, write factory, and read_time. The expression is somewhat complex and not easy to understand

**Comment by Tinet-mucw** [comment]
I think `code 1` is used to resolve the RTT issue;
my code is used to resolve the lost issue;
perhaps we can refer to the logic monitor jump in ast_read  of channel.c.


```shell
diff --git a/asterisk/main/audiohook.c b/asterisk/main/audiohook.c
index ac17566fe..7f20d32d6 100644
--- a/asterisk/main/audiohook.c
+++ b/asterisk/main/audiohook.c
@@ -268,13 +268,13 @@ static struct ast_frame *audiohook_read_frame_both(struct ast_audiohook *audioho
        }
 
        /* If we want to provide only a read factory make sure we aren't waiting for other audio */
-       if (usable_read && !usable_write && (ast_tvdiff_ms(ast_tvnow(), audiohook->write_time) < (samples/8)*2)) {
+       if (usable_read && !usable_write && (ast_tvdiff_ms(ast_tvnow(), audiohook->write_time) < (samples/8)*2) && (ast_slinfactory_available(&audiohook->read_factory) < 4 * samples)) {
                ast_debug(3, "Write factory %p was pretty quick last time, waiting for them.\n", &audiohook->write_factory);
                return NULL;
        }
 
        /* If we want to provide only a write factory make sure we aren't waiting for other audio */
-       if (usable_write && !usable_read && (ast_tvdiff_ms(ast_tvnow(), audiohook->read_time) < (samples/8)*2)) {
+       if (usable_write && !usable_read && (ast_tvdiff_ms(ast_tvnow(), audiohook->read_time) < (samples/8)*2) && (ast_slinfactory_available(&audiohook->write_factory) < 4 * samples)) {
                ast_debug(3, "Read factory %p was pretty quick last time, waiting for them.\n", &audiohook->read_factory);
                return NULL;
        }
```

File: main/audiohook.c [hunk 1]
```diff
@@ -281,13 +281,13 @@ static struct ast_frame *audiohook_read_frame_both(struct ast_audiohook *audioho
 	}
 
 	/* If we want to provide only a read factory make sure we aren't waiting for other audio */
```
Comment (jcolp) [inline main/audiohook.c:283]: This logic has grown complicated enough that I think comments should be added to each explaining the logic and why it is present and how it works

**Review by jcolp — CHANGES_REQUESTED** [review]

**Comment by Tinet-mucw** [comment]
I would greatly appreciate it if someone could help me standardize the comments.

File: main/audiohook.c [hunk 2]
```diff
@@ -280,14 +280,38 @@ static struct ast_frame *audiohook_read_frame_both(struct ast_audiohook *audioho
 		return NULL;
 	}
 
-	/* If we want to provide only a read factory make sure we aren't waiting for other audio */
-	if (usable_read && !usable_write && (ast_tvdiff_ms(ast_tvnow(), audiohook->write_time) < (samples/8)*2)) {
+	/* usable_read=1: indicates that read factory have the required samples
+	 * usable_write=0: indicates that write factory have not the required samples
+	 * The reason for useable_write=0 as fllows:
```
Comment (jcolp) [inline main/audiohook.c:285]: usable_write, follows

**Review by jcolp — CHANGES_REQUESTED** [review]

**Review by jcolp — APPROVED** [review]

**Comment by asteriskteam** [comment]
This PR has been marked stale because it has been in "Changes Requested" or "submitter-action-required" state for 28 days or more. Please make the requested changes within 14 days or the PR will be closed.

**Pull request merged**